### DISCUSSION TITLE
Gnome-terminal in codespaces

### DISCUSSION THREAD

**Discussion by Toshiro-kyun** [community/community#86709] [Codespaces] [answered]
### Select Topic Area

Question

### Body

Heyo, I want to use a shell script to open a new terminal and run a new command on that terminal (in codespaces). I'm trying to use gnome-terminal, but i get the following error:

"Unable to init server: Could not connect: Connection refused
#Failed to parse arguments: Cannot open display: "

How do I fix this? Help would highly be appreciated, I'm new to coding :). Thank you in advance

**Comment by georgetoloraia** [comment, marked as answer]
1. **Check if you are in a graphical environment:** 
- First, ensure that you are running your script in an environment that supports a graphical display. If you are using GitHub Codespaces or a similar cloud-based development environment, it's likely that it doesn't support GUI-based applications like `gnome-terminal` out of the box.

2. **Using `xterm` or `screen` instead:**
- In headless environments (like most remote servers or Codespaces), you might not have a graphical display. In such cases, consider using terminal multiplexers like `screen` or `tmux`, or try `xterm` if you really need a new terminal window. These tools are more suited for command-line environments and do not require a GUI.

3. **SSH X11 Forwarding:**
- If you are accessing the server (where the script runs) via SSH, you can use X11 forwarding to allow graphical applications to display on your local machine. You can enable this by connecting with the `-X` option (`ssh -X username@hostname`). However, this is only applicable if you are connecting to a remote server via SSH and have a local X server running on your machine.

4. **Check DISPLAY variable:**
- If you are sure that you are in a graphical environment and still face the issue, check if the `DISPLAY` environment variable is set correctly. You can print it in your shell script before you run `gnome-terminal`:
```
echo $DISPLAY
gnome-terminal
```
The `DISPLAY` variable should typically be something like `:0`. If it's not set, you might need to set it manually:
```
export DISPLAY=:0
gnome-terminal
```

5. **Allowing X11 Connection:**
- If the `DISPLAY` variable is set correctly, the X server might not be allowing connections from your application. You can allow connections from all clients by using `xhost +` but be aware that it's insecure to leave it like that. It's better to allow connections only from the needed client.

6. **Use `nohup` and `&`:**
- If you are just looking to run a script or command in the background from a terminal and don't necessarily need to interact with it, you might not need `gnome-terminal`. You can use `nohup` and `&` to run your command in the background:
```
nohup your-command &
```
This will run your-command in the background, and it will keep running even if the terminal is closed.
